### ISSUE TITLE
Multi narrator.

### ISSUE THREAD

**Issue by deam0n** [GeiserX/jelly-subtitles#151]
### What would you like?

Allow to detect different narrators / actors and separate the subtitle per narrator..

currently it creates subtitles all in the same line without ponctuation or séparation for different narrators 

### Alternatives considered

nothing I could think of

**Comment by GeiserX** [comment]
Thank you for taking the time to write this up, and for describing both symptoms rather than just one — they turn out to be two separate problems with very different answers, and one of them you can fix today in about thirty seconds.

## The run-on lines — fixable right now

Let me start here, because I suspect this is the part that actually spoils your viewing.

whisper.cpp has a `--max-len` option controlling the maximum segment length in characters, and **its default is `0`, meaning unlimited**. This plugin doesn't override that default, so you get whatever long unbroken segment whisper decides to emit. That is exactly your "all in the same line".

Go to **Dashboard → Plugins → WhisperSubs → Custom Whisper Arguments** and enter:

```
--max-len 47 --split-on-word
```

`--max-len 47` caps each cue near the ~42–47 characters per line that broadcast subtitling uses, and `--split-on-word` makes the break fall on a word boundary instead of mid-word. (That exact string is already the placeholder in the field.) This applies to the local whisper-cli engine. Existing subtitles are unchanged, so regenerate an item to see the difference.

## Speaker separation — a real limitation, not a missing toggle

This one I have to be straight with you about.

Whisper is a speech-**to-text** model. It was trained to predict text tokens from audio, and speaker identity never appears in its training targets or its output vocabulary. There is no speaker information computed anywhere inside it for the plugin to expose. So this isn't a setting I've neglected to add — the data doesn't exist in the model's output.

whisper.cpp ships two options that sound like they solve this, and it's worth explaining why neither does:

**`-di` / `--diarize`** is documented as "stereo audio diarization", and the implementation is literally a loudness comparison between the left and right channels: if the left channel is >1.1× louder it says speaker 0, if the right is louder it says speaker 1, and otherwise it emits `?`. It's gated on the input having exactly 2 channels. It was designed for recordings where each speaker is hard-panned to their own channel — a two-mic interview, a phone call. In a film or TV mix, dialogue is mixed to the centre and arrives at near-identical energy in both channels, so essentially every line falls into the `?` branch. It doesn't recognise voices at all.

**`-tdrz` / `--tinydiarize`** is genuine speaker-turn detection, but the constraints are severe. It requires the specific `ggml-small.en-tdrz.bin` checkpoint, which is **English-only** (finetuned on the AMI meeting corpus) and a **small** model. This plugin downloads the `large-v3` family precisely because accuracy matters for film dialogue. Enabling tinydiarize would mean every non-English user loses transcription entirely and English users get materially worse text. And it still wouldn't give you what you asked for: it emits a bare `[SPEAKER_TURN]` marker meaning "somebody else started talking" — not who, not how many people. It's explicitly experimental, with ~71% turn recall on its own benchmark.

## What real diarization would actually take

Proper diarization (WhisperX, `whisper-diarization`, NVIDIA NeMo) is a second machine-learning pipeline running alongside Whisper: voice-activity detection → neural speaker embeddings → clustering → aligning that speaker timeline onto Whisper's word timestamps. Concretely that means a Python and PyTorch runtime next to this plugin's self-contained native binary, plus pyannote's gated models, which require a Hugging Face account and manually accepting a licence before they can be downloaded at all. It is typically the slowest and most memory-hungry stage of the pipelines that use it — often comparable to the transcription itself.

The harder problem is that it wouldn't work well on the content you want it for. Published error rates for the current pyannote pipeline are about 18.8% DER on AMI meeting audio and 21.7% on DIHARD III. A peer-reviewed study specifically on TV series found audio-only diarization at **25.6% DER**, and attributed the difficulty to exactly what film audio is made of: overlapping dialogue, background score and effects masking voice differences, and short speaker turns. Roughly a quarter of speech mislabelled is not something I could ship and call correct.

And even at its best it produces anonymous cluster labels — `SPEAKER_00`, `SPEAKER_01` — never character or actor names. Getting from those to "Ellie" and "Joel" is a further unsolved step requiring per-show voice enrollment.

For completeness on the formatting question: `.srt` has no speaker field either, so even perfect diarization has to be rendered by convention — Netflix's guide uses a leading hyphen per speaker, one speaker per line, with a bracketed name when the speaker isn't visible on screen.

## One honest exception

I don't want to overstate the impossibility. It's true of the **local** engine, but this plugin can also talk to remote OpenAI-compatible transcription endpoints, and OpenAI now offers a dedicated `gpt-4o-transcribe-diarize` model that does diarization server-side, returning speaker-labelled output via a `diarized_json` response format (labels come back as "A", "B", … unless you supply reference samples).

That is a genuine path to what you're asking for — but I'm not claiming it as a feature, because the plugin currently negotiates `srt` / `verbose_json` and neither requests nor parses `diarized_json`, so pointing the remote URL at it today will not work. It's also a paid cloud service that uploads your audio, which is the opposite of why most people run this plugin. I'm noting it so you know the option exists in the wider world.

## On punctuation

Whisper normally does punctuate, so if you're getting none at all, the usual causes are: a small/base/tiny model (use the built-in downloader — `Large V3 Turbo (Q5)` is the recommended default at 574 MB), or a wrong language guess, which applies one language's punctuation conventions to another's words. If you know the language, set it explicitly instead of leaving it on auto. You can also try `--beam-size 8` in the same Custom Whisper Arguments field (the default is 5). VAD and non-speech suppression are already on by default.

---

I'm going to close this as not planned, because the speaker-separation half is blocked by what the underlying model genuinely doesn't compute rather than by work I'm declining to do — and I'd rather tell you that than leave the issue open implying it's coming.

But please do try `--max-len 47 --split-on-word` first. I think it addresses the readability problem that probably prompted you to open this. If punctuation is still missing afterwards, open a fresh issue with the model you're using, the audio language, and a few lines of the output, and I'll happily dig into that specifically.

Thanks again for the report — the run-on line default is a fair thing to have flagged.

**Comment by GeiserX** [comment]
Quick follow-up, since your report led to a real change.

Digging into the "all in the same line" half of this, I found it wasn't just whisper being whisper — it was a default on our side. whisper.cpp's `--max-len` (maximum characters per subtitle cue) defaults to `0`, meaning unlimited, and the plugin never overrode it. So whenever the model failed to punctuate a stretch of speech, there was no cap to stop the whole thing landing in one enormous cue.

So I've added a **Maximum subtitle line length** setting (Dashboard → Plugins → WhisperSubs, under the advanced section). It passes `--max-len` along with `--split-on-word`, so a capped line breaks on a word instead of mid-word. `47` is a reasonable starting value — broadcast subtitling caps a line near 42 characters. It defaults to `0` (off), so nobody's existing subtitles change on upgrade. If you run remote workers, the worker image gained the matching `WHISPER_MAX_LEN` knob, since a worker segments its own output.

You don't have to wait for that release, though — the same thing works today by putting `--max-len 47 --split-on-word` into the **Custom Whisper Arguments** field, as in my earlier reply.

Worth repeating one caveat: this is a guard against the worst case, not a cure for missing punctuation. If punctuation is absent entirely, the cause is usually a small model or a wrong language guess — so it's still worth checking you're on a `large-v3` model and, if you know the audio language, setting it explicitly rather than leaving it on auto.

The speaker-separation part of your request is unchanged — that remains out of scope for the reasons above.

Thanks again for taking the time to report this. The multi-narrator ask turned out not to be something I can build, but the second thing you mentioned almost in passing was a genuine defect that affected everyone, and it wouldn't have been found without your report.
